### PR TITLE
Fix #265: Index range bugs in YamlHeaderParser

### LINKED ISSUE #265: First character of text is swallowed in "extra" markdown
Using docfx 1.7.1 on Windows, with dnx.

docfx.json:

```json
{
  "metadata": [ { "src": [{ "files": "project.json" }], "dest": "obj/api" } ],
  "build": {
    "content": [{
      "files": [ "**/*.yml" ],
      "cwd": "obj/api",
      "dest": "api"
    }],
    "overwrite": "extra.md",
    "dest": "_site"
  }
}
```

project.json:

```json
{
  "version": "1.0.0",
  "frameworks": { "net451": { } }
}
```

DocfxTest.cs:

```csharp
namespace Foo
{
  /// <summary>Just a demo</summary>
  public class Bar {}
}
```

extra.md - not the lack of a blank line after the second "---"

```md
---
uid: Foo.Bar
---
abc
```

The resulting HTML file for `Foo.Bar` contains:

```html
              <div class="markdown level0 summary"><p>Just a demo</p>
</div>
              <div class="markdown level0 conceptual"><p>bc</p>
</div>
```

Note how the `a` of `abc` has been removed.

### PR DESCRIPTION
**Pull request by qinezh**
@vicancy @chenkennt @superyyrrzz @vwxyzh

### CODE REVIEW

File: test/Microsoft.DocAsCode.Build.Common.Tests/MarkdownReaderTest.cs [hunk 1]
```diff
@@ -13,21 +13,41 @@ public class MarkdownReaderTest
...
             Assert.NotNull(results);
             Assert.Equal(1, results.Count);
             Assert.Equal("Test", results[0].Uid);
             Assert.Equal("Hello", results[0].Metadata["remarks"]);
-            Assert.Equal(@"
-This is unit test!", results[0].Conceptual);
+            Assert.Equal(@"This is unit test!", results[0].Conceptual);
+            File.Delete(FileName);
+
+            // Test conceptual content between two yamlheader
```
Comment (superyyrrzz) [inline test/Microsoft.DocAsCode.Build.Common.Tests/MarkdownReaderTest.cs]: put the second case to a new method?

Comment (qinezh) [inline test/Microsoft.DocAsCode.Build.Common.Tests/MarkdownReaderTest.cs]: To be consist with `YamlHeaderParserUnitTest.cs`, merging several test case into one method...

Comment (qinezh) [inline test/Microsoft.DocAsCode.Build.Common.Tests/MarkdownReaderTest.cs]: To be consist with `YamlHeaderParserUnitTest.cs`, merging several test case into one method... 

---
In reply to: [61986408](https://github.com/dotnet/docfx/pull/285#discussion_r61986408) [](ancestors = 61986408)

Comment (superyyrrzz) [inline test/Microsoft.DocAsCode.Build.Common.Tests/MarkdownReaderTest.cs]: `YamlHeaderParserUnitTest.cs` also doesn't follow the [Arrange-Act-Assert](http://dotnet.github.io/docfx/guideline/csharp_coding_standards.html#unit-test-structure) pattern... Extract to a [theory] ?

---
In reply to: [61986933](https://github.com/dotnet/docfx/pull/285#discussion_r61986933) [](ancestors = 61986933,61986408)

File: src/Microsoft.DocAsCode.Build.Common/MarkdownReader.cs [hunk 2]
```diff
@@ -55,7 +55,7 @@ private static IEnumerable<OverwriteDocumentModel> ReadMarkDownCore(string file)
                     {
                         Uid = item.Id,
                         Metadata = item.Detail.Properties,
-                        Conceptual = content.Substring(start, end - start + 1),
+                        Conceptual = content.Substring(start, end - start).Trim(),
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Common/MarkdownReader.cs]: `Trim` is not correct. For Markdown syntax, trailing whitespaces has meanings.

File: src/Microsoft.DocAsCode.Build.Common/Coordinate.cs [hunk 3]
```diff
@@ -36,15 +35,9 @@ public static Coordinate GetCoordinate(string content)
                 return Default;
             }
             int index = content.Length - 1;
-            int line = content.Count(c => c == NewLineCharacter); // Assume there is no new line at the end of the file, count(line) = count(newline) + 1
+            int line = Regex.Matches(content, Environment.NewLine).Count;
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Common/Coordinate.cs]: using `Environment.NewLine` is not correct. Markdown content is not necessarily be `\r\n` under Windows.

Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Common/Coordinate.cs]: \r?\n

File: test/Microsoft.DocAsCode.Build.Common.Tests/MarkdownReaderTest.cs (hunk 1 above)
Comment (qinezh) [inline test/Microsoft.DocAsCode.Build.Common.Tests/MarkdownReaderTest.cs]: will refactor these Unit Test later

---
In reply to: [61987696](https://github.com/dotnet/docfx/pull/285#discussion_r61987696) [](ancestors = 61987696,61986933,61986408)

File: test/Microsoft.DocAsCode.Build.Common.Tests/MarkdownReaderTest.cs [hunk 4]
```diff
@@ -13,21 +13,52 @@ public class MarkdownReaderTest
...
-            File.WriteAllText(FileName, Content);
+            File.WriteAllText(FileName, content);
             var results = MarkdownReader.ReadMarkdownAsOverwrite(Environment.CurrentDirectory, FileName);
             Assert.NotNull(results);
             Assert.Equal(1, results.Count);
             Assert.Equal("Test", results[0].Uid);
             Assert.Equal("Hello", results[0].Metadata["remarks"]);
-            Assert.Equal(@"
-This is unit test!", results[0].Conceptual);
+            Assert.Equal("\r\nThis is unit test!", results[0].Conceptual);
```
Comment (vicancy) [inline test/Microsoft.DocAsCode.Build.Common.Tests/MarkdownReaderTest.cs]: use @"
" for assertion or use `\r\n` for content definition to keep consistency

File: test/Microsoft.DocAsCode.Build.Common.Tests/MarkdownReaderTest.cs [hunk 5]
```diff
@@ -13,21 +14,80 @@ public class MarkdownReaderTest
...
+            content = @"---
+uid: Test1
+remarks: Hello
+---
+This is unit test!
+---
+uid: Test2
+---
+";
+            content = Regex.Replace(content, "/r?/n", "/r/n");
```
Comment (vicancy) [inline test/Microsoft.DocAsCode.Build.Common.Tests/MarkdownReaderTest.cs]: can be extracted to a common method

**Comment by vicancy** [comment]
:+1:

**Pull request merged**